### PR TITLE
Fix for pre-bound snapshot empty source error

### PR DESCRIPTION
**Pull request by hakanmemisoglu**
* Use `GetSnapshotStatus` for pre-bound snapshots with nil `Source` field.

The pre-bound snapshots have  empty `Source` field in their `Spec`. Because of it, the call to `getCreateSnapshotInput` fails when the function tries to get the source volume. If we want to get the status for snapshot without `Source` field set, we can switch using `GetSnapshotStatus`.

### CODE REVIEW

File: pkg/controller/snapshot_controller.go [hunk 1]
```diff
@@ -560,16 +560,27 @@ func (ctrl *csiSnapshotController) checkandUpdateBoundSnapshotStatusOperation(sn
 	var timestamp int64
 	var size int64
 	var readyToUse = false
-	class, volume, _, snapshotterCredentials, err := ctrl.getCreateSnapshotInput(snapshot)
-	if err != nil {
-		return nil, fmt.Errorf("failed to get input parameters to create snapshot %s: %q", snapshot.Name, err)
-	}
-	driverName, snapshotID, timestamp, size, readyToUse, err := ctrl.handler.CreateSnapshot(snapshot, volume, class.Parameters, snapshotterCredentials)
```
Comment (tdmanv) [inline pkg/controller/snapshot_controller.go:567]: it's weird that they would predeclare some of these variables.

File: pkg/controller/snapshot_controller.go [hunk 2]
```diff
@@ -560,16 +560,27 @@ func (ctrl *csiSnapshotController) checkandUpdateBoundSnapshotStatusOperation(sn
...
+		readyToUse, timestamp, size, err = ctrl.handler.GetSnapshotStatus(content)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		class, volume, _, snapshotterCredentials, err := ctrl.getCreateSnapshotInput(snapshot)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get input parameters to create snapshot %s: %q", snapshot.Name, err)
+		}
+		var driverName string
```
Comment (tdmanv) [inline pkg/controller/snapshot_controller.go:575]: these are only used in the log?

**Comment by julio-lopez** [comment]
@hakanmemisoglu We are in the process of creating consistent guidelines for managing forked repos. I'll document those, but in short, we want to keep the `master` branch unmodified and use that branch to track the original repository's master branch. And use a `kasten` branch that is local to the forked repo as the branch with the kasten-specific changes.

The workflow would be:
* Develop in local branch
* Open PR in the Kasten's fork to merge changes to the `kasten` branch, review and merge.
* Open a separate PR (either from the `kasten` branch or a separate branch) to merge the changes upstream.

Periodically,
* The fork's `master` branch will be synced with the upstream `master` branch.
* The fork's `master` branch can be merged (via a PR) into the `kasten` branch to consume upstream changes.

I'll put the document together and tag you.

**Review by tdmanv — APPROVED** [review]

**Comment by pavannd1** [comment]
@hakanmemisoglu @julio-lopez
https://gist.github.com/Chaser324/ce0505fbed06b947d962 seems useful too.

File: pkg/controller/snapshot_controller.go (hunk 2 above)
Comment (hakanmemisoglu) [reply pkg/controller/snapshot_controller.go:575]: @tdmanv Yes, that's correct.

**Comment by hakanmemisoglu** [comment]
Closing this one... I will send a PR to upstream.

**Pull request merged**